ibm/resource_ibm_cdn.go

```diff
@@ -0,0 +1,579 @@
+		Update: resourceIBMCDNUpdate,
+		Delete: resourceIBMCDNDelete,
+		Exists: resourceIBMCDNExists,
+
+		Schema: map[string]*schema.Schema{
+			"hostname": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"vendor_name": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+			},
+
+			"origin_type": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  "HOST_SERVER",
+			},
+			"origin_address": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"bucketname": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"protocol": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  "HTTP",
+			},
+			"httpport": &schema.Schema{
+				Type:     schema.TypeInt,
+				Optional: true,
+				Default:  80,
+			},
+			"status": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"httpsport": &schema.Schema{
+				Type:     schema.TypeInt,
+				Optional: true,
+				Default:  443,
+			},
+			"cname": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  false,
+			},
+			"header": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  "",
+			},
+			"respectheaders": &schema.Schema{
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  true,
+			},
+			"fileextension": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  "",
+			},
+			"certificatetype": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  "0",
+			},
+			"cachekeyqueryrule": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  "include-all",
+			},
+			"performanceconfiguration": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  "General web delivery",
+			},
+			"path": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  "/",
+			},
+		},
+	}
+}
+
+func resourceIBMCDNCreate(d *schema.ResourceData, meta interface{}) error {
+	///create  session
+	sess := meta.(ClientSession).SoftLayerSession()
+	log.Println("ordering cdn service...")
+	///get the value of all the parameters
+	domain := d.Get("hostname").(string)
+	vendorname := d.Get("vendor_name").(string)
+	origintype := d.Get("origin_type").(string)
+	originaddress := d.Get("origin_address").(string)
+	protocol := d.Get("protocol").(string)
+	httpport := d.Get("httpport").(int)
+	httpsport := d.Get("httpsport").(int)
+	bucketname := d.Get("bucketname").(string)
+	path := d.Get("path").(string)
+	header := d.Get("header").(string)
+	cachekeyqueryrule := d.Get("cachekeyqueryrule").(string)
+	performanceconfiguration := d.Get("performanceconfiguration").(string)
+	respectheaders := d.Get("respectheaders").(bool)
+	cname := d.Get("cname").(string)
+	certificateType := d.Get("certificatetype").(string)
```

Comment:
For optional fields use d.GetOk line 130 to line 135 can be replaced 
var cname string
if name, ok := d.GetOk("cname"); ok {
   cname = name.(string) + str
}